### ISSUE TITLE
[speed] Can 1 robotic manipulator handle 2 cows simultaneously?

### ISSUE THREAD

**Issue by faerietree** [faerietree/autonomous_milking_ongoing_robot#1]
If the speed is high enough to milk two cows at once, then this allows a (less fail-safe/reliable) but cheaper robotic configuration with high throughput.

Other options are 2 manipulators for 2 cows, or even  4 for 2 cows - e.g. to allow chaining of cleaning and docking.

**Comment by klausz** [comment]
Is there any hardware defined for  the manipulator? Any sketches etc? Most important is a prototype. Then incremental improvements...

**Comment by faerietree** [comment]
The manipulator is undergoing a major redesign. Previously it had been a worm robot with joint rotation angle limited to 45 degree. Now it's a common 6 axes robotic manipulator.

Manipulator files are in the works. It'll be published in a separate repository and linked here as a submodule.

Do you have a proposal for the milking robot?

**Comment by klausz** [comment]
I agree, using a worm robot cannot be used currently. This concept will take another 5-10 years. 
My experience with robots is limited. There is experience with 3D Printer positioning (the Ultimaker 1 is very open and most reliable). Further a Mach3 controlled stepper CNC machine. Right now i am bringing a Robomow RL500 lawnmower back to life using the ardumower.de components. Also I am watching the franka.de 10000€ robot which is available in 2017.
We have the user requirements. To get good ideas it would be best to go on a makerfaire with the relevant cow and melking components. There are so many people who could give ideas and even join the project for that i suppose has to be in a final prive range of around 5000 US$ (max 10000 US$) so that it can be used (and maybe buildt) locally even from the 5-10 cows farmers. It would give them some freedom not to melk 2 times a day themselves.
For the concept, lets consider a central arm. On the arm one (maybe already all 4 units) are positioned below the cow. As it is in the middle (below the cow) it can rotate by 360 degrees and the length of the arm is used to position it (a simple polar coordinate system). Similar to what is used in a 3D Printer. When the given position (from camera or even simple position indicators) is reached it will go up. When it is working the vacuum can be measured.
Re positioning have a look at the ardumover.de They described an accuracy in the garden of 2cm. Putting 1-4 units on a cow (we have to be very careful if this is possible and should be done but it should be investigated) in a very low range are should deliver a much higher accuracy.
As you can see, there are only some ideas.

**Comment by klausz** [comment]
To start more easy instead of the robot, a manipulator could be used thats remote controlled from a human. Step by step is habdling system will be improved to a closed loop robotic system. 
(also it could make sense to develop a simple device now that prepares and checks the milk so that it can be sold officially dorectly from the farm - for Germany possibly a short heat up process).

**Comment by faerietree** [comment]
Thanks for the franka robot link. Promising, only pitfall it's not open source, is it?

Oliver and me at open source ecology already considered Ultimaker, which looks epic. Then we decided to go for an all belt design for speed. Thereafter, after considering to include metal working, it became apparent that the only _real_ solution (flexible in workspace & materials, saving resources compared to cartesian systems, low space consumption in home position) is a universal robot - not least due to the limited effort of open source and the huge amount of complicated tasks the so encouraging open source movement still needs solutions for urgently.

The price you mention (< $5000) is what we target. With the $200 Jetson ARM high processing computer cost is reduced drastically.

Also the AMOR design is scalable and reliable, the amount of animals being milked in parallel is free to choose.

As of the makerfaire you are right, yet for me it's impossible currently because I'm only working on all the open source projects in my free time. Budget also is so extremely tight that many projects are held back because the parts can't be ordered. (Trying to order the parts for the 6 axes manipulator since several weeks - we are ahead with development there a bit.) (This year every month the milk price falls by 1 euro cent. Nevertheless colleagues of mine recently started to help me save some time. When the foundation of worlddevelopment e.V. is official, the situation will improve further, because we'll be able to raise funds using Amazon Smile or via direct donations.)

Last but not least, the very nature of farmers, which are hardening and fixed at their view, prefer laughing at us rather than help. Also only few farmers are not self-focused, don't fear for their existence, nowadays more than ever with the ever falling milk prices - ecological farmers fear their market being flooded with conventional milk soon - hence ecological nearby farms eye the milking robot project with fear rather than hope, because they fear even more milk. I understand farmers, because it's never been any different throughout history. And I don't care if they bash me and my projects or not, because I'm not doing that all just for farmers or for just Europe where there's plenty of milk, I'm doing it for the good hearted people in the entire world - no matter where they are, I want to break up dependencies, dead locks and children poverty and fears.

> On the arm one (maybe already all 4 units) are positioned below the cow. As it is in the middle (below the cow) it can rotate by 360 degrees and the length of the arm is used to position it (a simple polar coordinate system). Similar to what is used in a 3D Printer. When the given position (from camera or even simple position indicators) is reached it will go up.

It was the initial plan to attach 4 teat cups at once. Then, from milking experience and a technical complexity evaluation it was followed that a one by one attachment does make more sense. Udders are too different in shape, spacing.

Point to discussion is whether saving significant docking time by omitting the for cows 5th (pre-milk, cleaning) cup is sensible or not. It will require 4 valves more because the pre milk must not be mixed with the normal milk. A solution may be to reuse the valves that are meant to redirect the milk to the animal kid (beast) milk containers. Id est the (pre-/beast-)milk must be evaluated there.

> Re positioning have a look at the ardumover.de They described an accuracy in the garden of 2cm. Putting 1-4 units on a cow (we have to be very careful if this is possible and should be done but it should be investigated) in a very low range are should deliver a much higher accuracy.
> Great ideas, comrade!

An alternative may be to use a follower algorithm combined with 3D vision which the robot will have anyways. The robot knows which cow is which because every cow needs are sender band (e.g. at the ankle). It then can approximate the cow's position due to the sender, then switch to 3D vision and use a follower algorithm to follow the udder and milk. This robot needs a rigid protective frame that doesn't hurt cows!
The automatic mobile calf-like milking process may also be in automatic mode, where the robot either checks for calf sender clustering around a cow nearby or visually checks for this. A problem could be that till the calf robot reaches the cow being milked by the real calf, the cow may already have lost patience. Also the cow may not allow a "foreign" calf to drink.

So the first version surely be the less nature/animal friendly variant: giving high energy food in a static yet movable (mobile from meadow to meadow), being equipped with 1 to 2 six axes robotic manipulators.

**Comment by faerietree** [comment]
The pasteurization is planned indeed, because I'd love to be able to directly put the milk to good with a cooking robot that creates fancy tarts like Dumbledore, Gandalf or @mention WhiteRaven89 can create them.

You find the first AMOR hardware specification in the manipulator repository: https://github.com/faerietree/manipulator/
